README.md
# drupal-emacs #


A Compilation of php-mode, drupal-mode and a number of other useful emacs modules. 

## Getting started ##

### Load module ###

When you clone this project remember to run:

git submodule update --init

If you place the module outside of emacs default load-path, then you must add it to the path manually in emacs init file:

> The init file must be in either of the following location on Linux,UNIX and OSX systems
> ~/.emacs, ~/.emacs.el, or ~/.emacs.d/init.el
> For Windows users look here [Emacs on Window FAQ](http://www.gnu.org/software/emacs/windows/).
    

Add this to the init file

    (add-to-list 'load-path "<SOME PATH>/drupal-emacs")
    e.g
    (add-to-list 'load-path "~/lib/drupal-emacs")

### Activate module ###

    (require 'drupal-emacs)
    (drupal-emacs)
### Performance ###

If you want emacs to load faster you should probably byte-compile the drupal-emacs directory. It is done like this:

    C-u 0 M-x byte-recompile-directory RET
    Navigate to drupal-emacs and RET again.

Remember to recompile when there are updates in the module(from git or custom).

## A way of using drupal-emacs with ctags ##

create a file called .dir-locals.el in the directory that you wish to be the root of your project at give it the following content

((nil . ((eval . (setq project-root-dir (locate-dominating-file buffer-file-name ".dir-locals.el"))))))

This will make emacs create a TAGS file in that directory and update it each time you save a file

## Quirks ##

c-subword-mode migt be void when the drupal-mode is loaded, this is coursed by an obsolite elc file(compiled emacs elisp file), most likely delivered by your GNU/Linux/*nix distribution.

You can test this by adding the following some where in init.el
 
    (add-hook 'php-mode-hook '(lambda ()(c-subword-mode t)))

If when you open a php file and the mini buffer says:

File mode specification error: (void-function c-subword-mode)

Then you need to update the cc-mode of your emacs.

### Howto ###

Become root

Change directory to /user/share

And call:

    find . -name "*cc-*.elc" -exec mv {} {}.old \;

Then download cc-mode from http://cc-mode.sourceforge.net/release.php

and extract it in the emacs lisp folder

To find that folder one option is to call

    find . -name "*cc-*.elc" (in /usr/share or /)


## Extra features ##


### MySQL screen client ###

The sh folder contains a MySQL pager script, which makes it possible to use gnu screen as a MySQL result browser.

#### How to ####

Copy or soft link my-screen-pager into path

start screen

    $ screen

split window

    C-a S

change window

    C-a <TAB>

create a new window

    C-a c

start a new screen session in this window

    $ screen -m

get and copy the session id

    $echo $STY

switch window

    C-a <TAB>

start mysql

    $mysql -u....

load script

    pager my-screen-pager <COPIED SESSION ID>

Now the bottom half of the window acts as a result browser, where you can search in the result and switch between the current a previous results.

The wrapper script pipes the result through less. So all the features of less are available in the result window.
The script uses a temp file for piping to less. The name of this file is echoed back to Mysql, and can be used to process the result in grep or anything else you please.



NEWS
2013-05-29  Kristian Nygaard Jensen  <freeduck@member.fsf.org>
	* This state of drupal-emacs is considered as BETA-1
	* It is now possible load the drupal-emacs module as any other emacs module


drupal-emacs.el
;; Copyright 2013 Kristian Nygaard Jensen

;; This file is part of drupal-emacs.

;; drupal-emacs is free software: you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; drupal-emacs is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with drupal-emacs.  If not, see <http://www.gnu.org/licenses/>.
(let ((default-directory (file-name-directory load-file-name)))
  (add-to-list 'load-path (file-truename "./elisp"))
  (add-to-list 'load-path (file-truename "./elisp/bookmark-plus"))
  (add-to-list 'load-path (file-truename "./elisp/geben"))
  (add-to-list 'load-path (file-truename "./elisp/icicles"))
  (add-to-list 'load-path (file-truename "./elisp/php-mode"))
  )

(defun drupal-emacs()
  (require 'icicles)
  (require 'dired+)
  (require 'bookmark+)
  
  (require 'php-mode)
  (require 'drupal-mode)

  (icy-mode 1)
  (load-theme 'manoj-dark t)
  (show-paren-mode t)

  (fset 'yes-or-no-p 'y-or-n-p)           ;replace y-e-s by y
  (setq x-select-enable-primary nil)
  (setq x-select-enable-clipboard t)
  (setq visible-bell t)
  (setq tags-revert-without-query t)
  (setq-default indent-tabs-mode nil)
  (autoload 'drupal-mode "drupal-mode" "Major mode for editing DRUPAL php code." t)

  (add-to-list 'auto-mode-alist '("\\.php$" . drupal-mode))
  (add-to-list 'auto-mode-alist '("\\.inc$" . drupal-mode))
  (add-to-list 'auto-mode-alist '("\\.module$" . drupal-mode))
  (add-to-list 'auto-mode-alist '("\\.install$" . drupal-mode))
  (add-to-list 'auto-mode-alist '("\\.test$" . drupal-mode))
  (add-to-list 'auto-mode-alist '("\\.txt$" . markdown-mode))
  (add-to-list 'auto-mode-alist '("\\.md$" . markdown-mode))
  (add-to-list 'auto-mode-alist '("\\.MD$" . markdown-mode))

  (global-set-key (kbd "<f9>") 'bmkp-file-all-tags-jump)

  (recentf-mode 1)
  (setq recentf-max-menu-items 25)
  (global-set-key "\C-x\ \C-r" 'recentf-open-files)

  (add-hook 'markdown-mode-hook 'select-markdown-theme)
  )

(defun select-markdown-theme ()
  "Set the prefered theme for markdown editing"
  (setq-default fill-column 80)
  (flyspell-mode t)
  )

(provide 'drupal-emacs)


elisp/drupal-mode.el
;; Copyright 2006-2012 Alexander I. Grafov,Daniel L. Gregoire, Ameen Ross, Kristian Nygaard Jensen

;; This file is part of drupal-emacs.

;; drupal-emacs is free software: you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; drupal-emacs is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with drupal-emacs.  If not, see <http://www.gnu.org/licenses/>.
;;;###autoload
(define-derived-mode drupal-mode php-mode "Drupal"
  "Major mode for DRUPAL coding.\n\n\\{drupal-mode-map}"
  (setq tab-width 2)
  (c-set-offset 'arglist-intro '+) ; for FAPI arrays and DBTNG
  (c-set-offset 'arglist-close '0)
  (setq c-basic-offset 2)
  (setq x-stretch-cursor t)
  (setq indent-tabs-mode nil)
  (setq fill-column 78)
  (setq show-trailing-whitespace t)
  (global-set-key (kbd "<f5>") "grep -n -R --exclude-dir=.svn --exclude=*~ --exclude=#* --exclude=TAGS --exclude=*emconf* -i -e bootstrap . 2>")
  (global-set-key (kbd "<f6>") "$this->")
  (global-set-key (kbd "<f7>") "DIRECTORY_SEPARATOR")
  (global-set-key (kbd "<f12>") 'imenu-add-menubar-index)
  (global-set-key (kbd "<f8>") "__METHOD__.__FILE__.__LINE__")
  (add-hook 'after-save-hook 'compile-tags)
  (add-hook 'before-save-hook 'delete-trailing-whitespace)
  (if (fboundp 'c-subword-mode)
      (c-subword-mode t))
  (run-mode-hooks 'drupal-mode-hook))

(provide 'drupal-mode)

(defun refresh-tags-table(&rest test)
  (let ((tag-file (concat project-root-dir "TAGS")))
    (visit-tags-table tag-file)))

(defun compile-tags-action (project-root-dir, current-dir)
  "compile etags for the current project"
  (cd project-root-dir)
  ;(start-file-process-shell-command "shell-update-tags" nil "rm TAGS;find -L .  -name \"*.php\" -o -name \"*.inc\" -o -name \"*.module\"|xargs ctags -ea --language-force=PHP")
  (start-file-process-shell-command "shell-update-tags" nil "rm TAGS;find  .  -name \"*.php\" -o -name \"*.inc\" -o -name \"*.module\"|xargs ctags -ea --language-force=PHP")
  (set-process-filter (get-process "shell-update-tags") 'refresh-tags-table)
  (cd current-dir))

(defun compile-tags ()
  "compile etags for the current project"
  (interactive)
  (if (boundp 'project-root-dir)
    (compile-tags-action project-root-dir (file-name-directory  buffer-file-name))))


init.el


;; Copyright 2012 Kristian Nygaard Jensen

;; This file is part of drupal-emacs.

;; drupal-emacs is free software: you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; drupal-emacs is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with drupal-emacs.  If not, see <http://www.gnu.org/licenses/>.
(add-to-list 'load-path "~/lib/drupal-emacs/elisp")
(add-to-list 'load-path "~/lib/drupal-emacs/elisp/php-mode")
(add-to-list 'load-path "~/lib/drupal-emacs/elisp/geben")
(add-to-list 'load-path "~/lib/drupal-emacs/elisp/bookmark-plus")
(add-to-list 'load-path "~/lib/drupal-emacs/elisp/icicles")

;; (add-to-list 'load-path "~/lib/elisp/ecb-2.40/")

;; (load-file "~/lib/elisp/cedet-1.0.1/common/cedet.el")
;; (global-ede-mode 1)                      ; Enable the Project management system
;; (semantic-load-enable-code-helpers)      ; Enable prototype help and smart completion
;; (global-srecode-minor-mode 1)            ; Enable template insertion menu


;; (require 'ecb)
;; (require 'ecb-autoloads)
(require 'coffee-mode)
(require 'doremi)
(require 'doremi-cmd)
(require 'doremi-frm)

(load-theme 'manoj-dark t)
(show-paren-mode t)

(fset 'yes-or-no-p 'y-or-n-p)           ;replace y-e-s by y
(setq x-select-enable-primary nil)
(setq x-select-enable-clipboard t)
(setq visible-bell t)
(setq tags-revert-without-query t)
(setq-default indent-tabs-mode nil)
(autoload 'drupal-mode "drupal-mode" "Major mode for editing DRUPAL php code." t)
(require 'drupal-mode)
(require 'php-mode)
(require 'sql-completion)
(setq mysql-password "staalanden")
(require 'dired+)
(require 'bookmark+)
(require 'icicles)

(recentf-mode 1)
(setq recentf-max-menu-items 25)
(global-set-key "\C-x\ \C-r" 'recentf-open-files)

(autoload 'geben "geben" "PHP Debugger on Emacs" t)

(add-to-list 'auto-mode-alist '("\\.php$" . drupal-mode))
(add-to-list 'auto-mode-alist '("\\.inc$" . drupal-mode))
(add-to-list 'auto-mode-alist '("\\.module$" . drupal-mode))
(add-to-list 'auto-mode-alist '("\\.install$" . drupal-mode))
(add-to-list 'auto-mode-alist '("\\.test$" . drupal-mode))
(add-to-list 'auto-mode-alist '("\\.txt$" . markdown-mode))
(add-to-list 'auto-mode-alist '("\\.md$" . markdown-mode))
(add-to-list 'auto-mode-alist '("\\.MD$" . markdown-mode))

;; Put autosave files (ie #foo#) and backup files (ie foo~) in ~/.emacs-backups/.
(custom-set-variables
 ;; custom-set-variables was added by Custom.
 ;; If you edit it by hand, you could mess it up, so be careful.
 ;; Your init file should contain only one such instance.
 ;; If there is more than one, they won't work right.
 '(auto-save-file-name-transforms (quote ((".*" "~/.emacs-backups/autosaves/\\1" t))))
 '(backup-directory-alist (quote ((".*" . "~/.emacs-backups/backups/"))))
 '(bmkp-last-as-first-bookmark-file "~/.emacs.bmk")
 '(bmkp-prompt-for-tags-flag t)
 '(bookmark-version-control t)
 '(geben-pause-at-entry-line t)
 '(geben-session-enter-hook (quote (geben-session-redirect-init geben-session-context-init geben-session-breakpoint-init geben-session-source-init drupal-emacs-geben-session-init)))
 '(geben-show-breakpoints-debugging-only t)
 '(icicle-mode nil)
 '(icicle-saved-completion-sets (quote (("breaking" . "/home/kristian/emacs-projects/breaking.icy"))))
 '(js-indent-level 2)
 '(safe-local-variable-values (quote ((eval setq project-root-dir (locate-dominating-file buffer-file-name ".dir-locals.el")))))
 '(tool-bar-mode nil))

(setq remember-annotation-functions '(org-remember-annotation))
(setq remember-handler-functions '(org-remember-handler))
(add-hook 'remember-mode-hook 'org-remember-apply-template)

(global-set-key (kbd "<f9>") 'bmkp-file-all-tags-jump)

;; create the autosave dir if necessary, since emacs won't.
(make-directory "~/.emacs-backups/autosaves/" t)
(make-directory "~/.emacs-backups/backups/" t)

(add-hook 'markdown-mode-hook 'select-markdown-theme)

(defun select-markdown-theme ()
  "Set the prefered theme for markdown editing"
  (setq-default fill-column 80)
  (flyspell-mode t))

(defun drupal-emacs-geben-session-init (session)
  (let (result (geben-eval-expression "$_SERVER[\"REQUEST_URI\"]"))
    (message "Printing the result")
    (message result)))
;;(geben-eval-expression "$_SERVER[\"REQUEST_URI\"]")


(setq sql-interactive-mode-hook
      (lambda ()
	(define-key sql-interactive-mode-map "\t" 'comint-dynamic-complete)
	(sql-mysql-completion-init)))

;; Enable directory local variables with remote files. This facilitates both
;; the (dir-locals-set-class-variables ...)(dir-locals-set-directory-class ...)
;; and the dir-locals.el approaches.
(defadvice hack-dir-local-variables (around my-remote-dir-local-variables)
  "Allow directory local variables with remote files, by temporarily redefining
     `file-remote-p' to return nil unconditionally."
  (flet ((file-remote-p (&rest) nil))
    ad-do-it))
(ad-activate 'hack-dir-local-variables)

(setq org-todo-keywords '((sequence "TODO(t)" "IN PROGRESS(i!/!)" "WAIT(w@/!)"  "|" "DONE(d@)" "CANCELED(c@)")))

(defun fullscreen ()
  (interactive)
  (x-send-client-message nil 0 nil "_NET_WM_STATE" 32
	    		 '(2 "_NET_WM_STATE_FULLSCREEN" 0))
  (menu-bar-mode -1)
  (tool-bar-mode -1)
  (scroll-bar-mode -1))

(global-set-key (kbd "<f11>") 'fullscreen)




(custom-set-faces
 ;; custom-set-faces was added by Custom.
 ;; If you edit it by hand, you could mess it up, so be careful.
 ;; Your init file should contain only one such instance.
 ;; If there is more than one, they won't work right.
 )

(icy-mode 1)
